### PR TITLE
Backend revamp

### PR DESCRIPTION
**Pull request by Lexseal**
### Features Added/Issues Addressed
A complete solution for the backend

### Brief Description of Solution
1. Whenever something is fetched, localstore will cache the recipe so both online and custom recipes will be saved to the same place.
2. A recipe hash now consists of the name of the recipe and the hash of the steps separated by `$`
3. There are also several arrays on the localstore to keep track of custom recipes and favorite recipes. These arrays have keys that starts with `%`. They have a hashset as filter so no element will be duplicated
4. The search functionality is a crude fuzzy search using minimum edit distance between two strings. The substitution cost is estimated by the distance of the characters on a standard qwerty keyboard.

### Checklist
- [x] Added new tests (You know the code best to make new tests)
- [x] Resolved merge conflicts (Prefer merge over rebase)
- [ ] Status check passed (Linted and tested)
- [ ] Reviewed (At least one reviewer's approval)

### Caveat/Special Instructions for Team
Let me know if you are unclear how to call one of these functions. Most of the time, you will be using the `search recipe` function.

### CODE REVIEW

File: source/assets/scripts/backend.js [hunk 1]
```diff
@@ -0,0 +1,344 @@
+/** @module backend */
+
+import { recipe_data, keep_fields } from './recipe-data.js';
+
+const API_KEY = '486eca841c6a49b896486723439f9977';
```
Comment (Lexseal) [inline source/assets/scripts/backend.js:5]: separating out the constants

File: source/assets/scripts/backend.js [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+      ch = str.charCodeAt(i);
+      h1 = Math.imul(h1 ^ ch, 2654435761);
+      h2 = Math.imul(h2 ^ ch, 1597334677);
+    }
+    h1 = Math.imul(h1 ^ (h1>>>16), 2246822507) ^ Math.imul(h2 ^ (h2>>>13), 3266489909);
+    h2 = Math.imul(h2 ^ (h2>>>16), 2246822507) ^ Math.imul(h1 ^ (h1>>>13), 3266489909);
+    return (4294967296 * (2097151 & h2) + (h1>>>0)).toString();
+  };
+  
+  return recipe.name + '$' + cyrb53(JSON.stringify(recipe.steps));
```
Comment (Lexseal) [inline source/assets/scripts/backend.js:69]: note that there is a `$` between the name and the cyrb hash.
Ex. apple pie$12987198327

File: source/assets/scripts/backend.js [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+ * search a recipe by its name and keep only the parts we need according to the global variable keep_fields
+ * we then remap the field names by the rule fields_remap
+ * afterwards, it saves all filtered recipes to the localstore
+ * @param {string} name - name of the recipe 
+ * @param {Array<string>} intolerances - a list of intolerances. for a list of supported keywords, see 
+ *  https://spoonacular.com/food-api/docs#:~:text=of%20supported%20diets.-,intolerances,-string
+ *  Note that this functionality is pretty bad since if a recipe can be nut free without listing nuts as intolerance
+ * @return {Promise} returns a list of fetched recipe for testing purpose
+ */
+export async function fetch_recipe(name, intolerances) {
```
Comment (Lexseal) [inline source/assets/scripts/backend.js:35]: I think we are going to ignore the intolerances and do our own filtering by the ingredients

File: source/assets/scripts/backend.js [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+}
+
+/**
+ * search recipe in the localstore and as well as an option to auto fetch from the internet
+ * @param {string} name - name of the recipe you want to search 
+ * @param {boolean} online - whether you want to pull in online results (local search by default)
+ * @param {float} match_tolerance - max distance between two strings that the fuzzy search allows
+ * @return {Promise} a list of recipe_hash
+ */
+export async function search_recipe(name, online=false, match_tolerance=10) {
```
Comment (Lexseal) [inline source/assets/scripts/backend.js:297]: This is the one you want to use in search page

File: source/assets/scripts/backend.js [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+}
+
+/**
+ * suggest some recipes a user might be looking for
+ * @param {string} name - user input
+ * @param {float} match_tolerance - max distance between two strings that the fuzzy search allows
+ * @param {float} return_size - max length of the return array
+ * @return an array of recipe names that might be the thing you are searching for sorted by relavance
+ */
+export async function search_suggest(name, match_tolerance=15, return_size=5) {
```
Comment (Lexseal) [inline source/assets/scripts/backend.js:331]: This give you suggestions on what you might want to search

File: source/assets/scripts/recipe-searchPage.js [hunk 6]
```diff
@@ -8,6 +10,21 @@ async function init() {
   goAdd();
   goSettings();
 
+  // sample usage of the search_recipe function
+  const search_button = document.getElementById('search-button');
+  search_button.addEventListener('click', () => {
+    const input = document.getElementById('search-field').value;
+    backend.search_recipe(input).then(console.log);
```
Comment (Lexseal) [inline source/assets/scripts/recipe-searchPage.js:17]: sample usage

File: source/assets/styles/recipe-searchPage.css [hunk 7]
```diff
@@ -91,7 +91,7 @@ nav {
     display:flex
 }
 
-  form.search input[type=text] {
+  div.search input[type=text] {
```
Comment (Lexseal) [inline source/assets/styles/recipe-searchPage.css:94]: changed the form into `div` since submitting a `form` will refresh the page

**Review by Xubangbangg — APPROVED** [review]
looks good

**Review by joyqiao1 — APPROVED** [review]

**Pull request merged**